sos/collector/transports/saltstack.py

```diff
@@ -0,0 +1,173 @@
+        salt_run_cmd = ['salt-run', 'manage.status']
+        _check_for_salt_master(
+            salt_cmd, "The salt binary is not available on this system")
+
+        _check_for_salt_master(
+            salt_run_cmd, "The salt-run binary is not available on this "
+            "system to confirm if the salt master is running.")
+        return True
+
+    def _connect(self, password=None):
+        """Connect to the remote host using SaltStack Master.
+
+        This method will attempt to connect to the remote host using SaltStack
+        Master. If the connection fails, an exception will be raised.
+
+        If the connection is successful, the connection will be stored in the
+        self._connection attribute.
+        """
+        if not is_executable('salt'):
+            return False
+
+        try:
+            self._check_for_saltstack()
+        except ConnectionException:
+            self.log_error("Transport is not locally supported. ")
+            raise
+        self.log_info("Transport is locally supported and service running. ")
+        cmd = "echo Connected"
+        result = self.run_command(cmd, timeout=180)
+        return result['status'] == 0
+
+    def _disconnect(self):
+        return True
+
+    @property
+    def remote_exec(self):
+        """The remote execution command to use for this transport."""
+        salt_args = "--out json --static --no-color"
+        cmd = "salt %s %s cmd.shell " % (salt_args, self.address)
+        return (cmd)
+
+    def _retrieve_file(self, fname, dest):
+        """Retrieve a file from the remote host using saltstack
+
+        This method will retrieve a file from the remote host and access it
+        from the master's minion cachedir.
+
+        Parameters
+            fname       The path to the file on the remote host
+            dest        The path to the destination directory on the master
+
+        Returns
+            True if the file was retrieved, else False
+        """
+        file = os.path.basename(fname)
+        if not self.connected:
+            return False
+        master_minion_cachedir = "/var/cache/salt/master/minions/%s/files"
+        master_minion_cachedir %= self.address
+        cachedir_file = os.path.join(master_minion_cachedir, fname.lstrip('/'))
+        result = self._salt_retrieve_file(
+            self.address,
+            fname,
+            "Failed to collect file: %s from node: %s" % (fname, self.address)
+        )
+        if result['status'] == 0:
+            copy_result = shutil.move(cachedir_file, dest)
+            if copy_result:
+                self.log_info("File %s copied into dest: %s" % (file, dest))
+                return True
+            else:
+                self.log_debug(
+                    "File copy was not successful for: %s" % cachedir_file
+                )
+                return False
+        else:
+            self.log_debug("Error copying file: %s" % fname)
+            return False
```

Comment:
Logging is handled in the base `RemoteTransport().retrieve_file()` which calls `_retrieve_file()`, as well as retries. A new transport just needs `_retrieve_file()` to execute the command and return True/False depending on if the retrieval was successful.